File: src/system/pubsub.jl
module PubSub
using InteractiveUtils
const global Topics = Base.ImmutableDict(
	:clock => "/clock",
	:position => "/pose", 
	:velocity => "/vel")

_topics() = Topics
	
const global channels = Dict{String, Set{Channel}}()
@code_llvm _topics()

for (k, v) in Topics
	channels[v] = Set{Channel}()
end

function publish(topic::String, data::Any)
	if haskey(channels, topic)
		cs = channels[topic]
		println(length(cs))
		for c in cs
			put!(c, data)
		end
	end
end

function subscribe(topic::String)
	cs = get!(channels, topic, Set())
	channel = Channel(1)
	push!(cs, channel)
	return channel
end

function unsubscribe(topic::String, channel::Channel)
	if haskey(channels, topic)
		cs = get!(channels, topic, [])
		delete!(cs, channel)
	end
end

export Topics, publish, subscribe, unsubscribe

end

File: py/system/pubsub.py
from enum import Enum
import enum
from typing import Dict, Set
from system.utils import SingletonDecorator

class Topics(Enum):
	clock = "/clock"
	position = "/pose"
	velocity = "/vel"

class PubSub:
	def __init__(self) -> None:
		self.callbacks = dict()
		for t in enum(Topics):
			self.callbacks[t] = set()
	
	def publish(self, topic, data):
		if self.callbacks.has_key(topic):
			callbacks = self.callbacks[topic]
			for c in callbacks:
				c(data)
		else:
			print('Topic not found')

	def subscribe(self, topic, callback):
		callbacks = self.callbacks[topic]
		if not callbacks:
			callbacks = set()
		callbacks.add(callback)
		self.callbacks = callbacks

	def unsubscribe(self, topic, callback):
		callbacks = self.callbacks[topic]
		if callbacks:
			callbacks.remove(callback)
			self.callbacks = callbacks

PubSubInstance = SingletonDecorator(PubSub)